Exception to deserialize 'com.spotify.docker.client.messages.Event'

## Description

I'm trying to serialize events from Docker to send for a monitoring tool. I can serialized an event to json, but when I try to deserialize  arise an exception : `Exception in thread "main" com.fasterxml.jackson.databind.JsonMappingException: For input string: "2017-01-11T06:43:34.000+0000" (through reference chain: com.spotify.docker.client.messages.Event["time"])`

## How to reproduce
```
package br.cin.gfads.adalrsjr1.docker_monitor;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.spotify.docker.client.DefaultDockerClient;
import com.spotify.docker.client.DockerClient;
import com.spotify.docker.client.EventStream;
import com.spotify.docker.client.ObjectMapperProvider;
import com.spotify.docker.client.messages.Event;

public class App 
{
	private static final Logger log = LoggerFactory.getLogger(App.class);
	
    public static void main( String[] args ) throws Exception
    {
    	DockerClient docker = DefaultDockerClient.fromEnv().build();
    	DefaultDockerClient.builder()
    					   .uri("localhost:2376")
    					   .connectionPoolSize(1)
    	                                   .build();
    	
    	EventStream eventStream = docker.events();
    	while(eventStream.hasNext()) {
    		Event ev = eventStream.next();
    		
    		ObjectMapperProvider omp = new ObjectMapperProvider();
    		ObjectMapper mapper = omp.getContext(Event.class);
    		
    		String json = mapper.writerWithView(Event.class).writeValueAsString(ev);
    		
    		Event ev2 = mapper.getFactory().createParser(json).readValueAs(Event.class);
//    		Event ev2 = mapper.readerFor(Event.class).readValue(json); // also not working
    		log.info(ev2+"");
    	}
    	
    	docker.close();
    	
    }
}
```

## What do you expect
I expect something like this:

`Event{status=null, id=null, from=null, type=NETWORK, action=disconnect, actor=Actor{id=hvb6mpc8t4iu7j85vu27sb1n6, attributes={container=a2ceade7f529694723695256c5d04b135e5573f9b05c597448e1a8d171a22231, name=ingress, type=overlay}}, time=Wed Jan 11 03:52:33 GFT 2017, timeNano=1484117553094885100}`

## What happened instead
`Exception in thread "main" com.fasterxml.jackson.databind.JsonMappingException: For input string: "2017-01-11T06:43:34.000+0000" (through reference chain: com.spotify.docker.client.messages.Event["time"])`

## Software:

- `docker version`:
Client:
 Version:      1.13.0-rc5
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.0-rc5
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      linux/amd64
 Experimental: true
- Spotify's docker-client version: 7.0.0

## Full backtrace

```text
Exception in thread "main" com.fasterxml.jackson.databind.JsonMappingException: For input string: "2017-01-11T06:54:27.000+0000" (through reference chain: com.spotify.docker.client.messages.Event["time"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:210)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:177)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1475)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:463)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:377)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1100)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:294)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:131)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3674)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1996)
	at com.fasterxml.jackson.core.JsonParser.readValueAs(JsonParser.java:1511)
	at br.cin.gfads.adalrsjr1.docker_monitor.App.main(App.java:39)
Caused by: java.lang.NumberFormatException: For input string: "2017-01-11T06:54:27.000+0000"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Long.parseLong(Long.java:589)
	at java.lang.Long.parseLong(Long.java:631)
	at com.spotify.docker.client.jackson.UnixTimestampDeserializer.deserialize(UnixTimestampDeserializer.java:47)
	at com.spotify.docker.client.jackson.UnixTimestampDeserializer.deserialize(UnixTimestampDeserializer.java:36)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:520)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:461)
	... 8 more
```

--- a/src/main/java/com/spotify/docker/client/ObjectMapperProvider.java
+++ b/src/main/java/com/spotify/docker/client/ObjectMapperProvider.java
@@ -86,7 +86,7 @@ public class ObjectMapperProvider implements ContextResolver<ObjectMapper> {
     return OBJECT_MAPPER;
   }
 
-  static ObjectMapper objectMapper() {
+  public static ObjectMapper objectMapper() {
     return OBJECT_MAPPER;
   }
 

--- a/src/main/java/com/spotify/docker/client/jackson/UnixTimestampSerializer.java
+++ b/src/main/java/com/spotify/docker/client/jackson/UnixTimestampSerializer.java
@@ -1,0 +1,47 @@
+/*-
+ * -\-\-
+ * docker-client
+ * --
+ * Copyright (C) 2016 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.docker.client.jackson;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+
+import java.io.IOException;
+import java.util.Date;
+
+/**
+ * A {@link Date} serializer that outputs seconds since epoch.
+ * The inverse of {@link UnixTimestampDeserializer}.
+ */
+public class UnixTimestampSerializer extends JsonSerializer<Date> {
+
+  private UnixTimestampSerializer() {
+  }
+
+  @Override
+  public void serialize(final Date date,
+                        final JsonGenerator gen,
+                        final SerializerProvider serializers)
+      throws IOException, JsonProcessingException {
+    gen.writeNumber(date.getTime() / 1000);
+  }
+}

--- a/src/main/java/com/spotify/docker/client/messages/Event.java
+++ b/src/main/java/com/spotify/docker/client/messages/Event.java
@@ -28,9 +28,11 @@ import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonValue;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableMap;
 import com.spotify.docker.client.jackson.UnixTimestampDeserializer;
+import com.spotify.docker.client.jackson.UnixTimestampSerializer;
 
 import java.util.Date;
 import java.util.Map;
@@ -98,6 +100,7 @@ public abstract class Event {
 
   @JsonProperty("time")
   @JsonDeserialize(using = UnixTimestampDeserializer.class)
+  @JsonSerialize(using = UnixTimestampSerializer.class)
   public abstract Date time();
 
   @Nullable

--- a/src/test/java/com/spotify/docker/client/jackson/UnixTimestampSerializerTest.java
+++ b/src/test/java/com/spotify/docker/client/jackson/UnixTimestampSerializerTest.java
@@ -1,0 +1,60 @@
+/*-
+ * -\-\-
+ * docker-client
+ * --
+ * Copyright (C) 2016 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.docker.client.jackson;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.spotify.docker.client.ObjectMapperProvider;
+
+import java.util.Date;
+
+import org.junit.Test;
+
+public class UnixTimestampSerializerTest {
+
+  private static final ObjectMapper OBJECT_MAPPER =
+      new ObjectMapperProvider().getContext(UnixTimestampSerializerTest.class);
+
+  private static class TestClass {
+
+    @JsonProperty("date")
+    @JsonSerialize(using = UnixTimestampSerializer.class)
+    private Date date;
+
+    TestClass(final Date date) {
+      this.date = date;
+    }
+  }
+
+  @Test
+  public void testToString() throws Exception {
+    final long timestamp = 1487357474682L;
+    final String expectedJson = "{\"date\":1487357474}";
+    final TestClass testClass = new TestClass(new Date(timestamp));
+
+    final String json = OBJECT_MAPPER.writeValueAsString(testClass);
+    assertThat(json, equalTo(expectedJson));
+  }
+}

--- a/src/test/java/com/spotify/docker/client/messages/EventTest.java
+++ b/src/test/java/com/spotify/docker/client/messages/EventTest.java
@@ -1,0 +1,50 @@
+/*-
+ * -\-\-
+ * docker-client
+ * --
+ * Copyright (C) 2016 - 2017 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.docker.client.messages;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableMap;
+import com.spotify.docker.client.ObjectMapperProvider;
+
+import java.util.Date;
+
+import org.junit.Test;
+
+public class EventTest {
+
+  @Test
+  public void serializationRoundTripTest() throws Exception {
+    // Test serializing and deserializing the same Event instance works and preserves data
+    final Event event = Event.create("create", "foo", "nginx", Event.Type.CONTAINER, "create",
+        Event.Actor.create("bar", ImmutableMap.of("image", "nginx", "name", "docker-nginx")),
+        new Date(1487356000), 100L);
+
+    final ObjectMapper mapper = ObjectMapperProvider.objectMapper();
+
+    final String json = mapper.writeValueAsString(event);
+
+    final Event event2 = mapper.readValue(json, Event.class);
+    assertThat(event, equalTo(event2));
+  }
+}

